scintillometry/fourier/base.py

```diff
@@ -0,0 +1,273 @@
+            f = [0, 1, ...,   n/2-1,     -n/2, ..., -1] * sample_rate
+
+        if n is even, and
+
+            f = [0, 1, ..., (n-1)/2, -(n-1)/2, ..., -1] * sample_rate
+
+        if n is odd.
+
+        For real time-domain data, `numpy.fft.rfftfreq` is used, which returns
+
+            f = [0, 1, ...,     n/2-1,     n/2] * sample_rate
+
+        if n is even, and
+
+            f = [0, 1, ..., (n-1)/2-1, (n-1)/2] * sample_rate
+
+        if n is odd.
+
+        If ``self.sample_rate`` is `None`, output is unitless.
+
+        Returns
+        -------
+        freqs : `~numpy.ndarray`
+            Sample frequencies.
+        """
+        sample_rate = self.sample_rate
+        if sample_rate is None:
+            sample_rate = 1.
+        a_length = self.data_format['time_shape'][self.axis]
+        if 'float' in self.data_format['time_dtype'].name:
+            return np.fft.rfftfreq(a_length) * sample_rate
+        return np.fft.fftfreq(a_length) * sample_rate
+
+    def inverse(self):
+        return self._parent.inverse(self.data_format, self.direction,
+                                    self.axis, self.norm, self.sample_rate)
+
+    def __call__(self, a):
+        """Fast Fourier transform.
+
+        For the direction of the transform and shapes and dtypes of the arrays,
+        use repr().
+
+        Parameters
+        ----------
+        a : array_like
+            Input data.
+
+        Returns
+        -------
+        out : `~numpy.ndarray`
+            Transformed data.
+        """
+        return self._fft(a)
+
+    def __repr__(self):
+        return ("<{s.__class__.__name__} "
+                " engine={s._parent._engine_name},"
+                " direction={s.direction},\n"
+                "    axis={s.axis}, norm={s.norm},"
+                " sample_rate={s.sample_rate}\n"
+                "    Time domain: shape={fmt[time_shape]},"
+                " dtype={fmt[time_dtype]}\n"
+                "    Frequency domain: shape={fmt[freq_shape]},"
+                " dtype={fmt[freq_dtype]}>"
+                .format(s=self, fmt=self.data_format))
+
+
+class NumpyFFTMaker(FFTMakerBase):
+    """FFT factory class utilizing `numpy.fft` functions.
+
+    FFTs of real-valued time-domain data use `~numpy.fft.rfft` and its inverse.
+    `~numpy.fft.rfft` performs a real-input transform on one dimension of the
+    input, halving that dimension's length in the output.
+
+    Currently does not support Hermitian FFTs (`~numpy.fft.hfft`, etc.).
+    """
+
+    _engine_name = 'numpy'
+
+    def _setup_transform(self, data_format, direction, axis, norm,
+                         sample_rate):
+        """Set up `numpy.fft` based FFT using metadata."""
+
+        complex_data = 'complex' in data_format['time_dtype']
+
+        if direction == 'forward':
+            if complex_data:
+
+                def fft(a):
+                    return np.fft.fft(a, axis=axis, norm=norm).astype(
+                        data_format['freq_dtype'])
```

Comment:
Here and below, add `copy=False` in the `astype` (so that nothing gets done if the type is already OK).